File: problems/0002-add-two-numbers/solution.py
"""
Problem Number: 2
Problem Name: Add Two Numbers
Difficulty: Medium
Topics: Linked List, Math
Companies: Amazon, Google, Microsoft, Meta
"""

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Approach: Elementary Math
        Time Complexity: O(max(m, n))
        Space Complexity: O(max(m, n))
        """
        dummy = ListNode(0)
        curr = dummy
        carry = 0
        
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            # New digit
            val = val1 + val2 + carry
            carry = val // 10
            val = val % 10
            
            curr.next = ListNode(val)
            curr = curr.next
            
            # Update pointers
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            
        return dummy.next

# Helper to create list from array
def createList(arr):
    dummy = ListNode(0)
    curr = dummy
    for val in arr:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

# Helper to print list
def printList(node):
    output = []
    while node:
        output.append(node.val)
        node = node.next
    print(output)

# Test cases
if __name__ == "__main__":
    solution = Solution()
    l1 = createList([2,4,3])
    l2 = createList([5,6,4])
    printList(solution.addTwoNumbers(l1, l2)) # [7,0,8]

File: problems/0002-add-two-numbers/solution.js
/**
 * @problemNumber: 2
 * @problemName: Add Two Numbers
 * @difficulty: Medium
 * @topics: Linked List, Math
 * @companies: Amazon, Google, Microsoft, Meta
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * Approach: Elementary Math
 * Time Complexity: O(max(m, n))
 * Space Complexity: O(max(m, n))
 * 
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
function addTwoNumbers(l1, l2) {
    const dummy = new ListNode(0);
    let curr = dummy;
    let carry = 0;

    while (l1 !== null || l2 !== null || carry !== 0) {
        let sum = carry;

        if (l1 !== null) {
            sum += l1.val;
            l1 = l1.next;
        }

        if (l2 !== null) {
            sum += l2.val;
            l2 = l2.next;
        }

        carry = Math.floor(sum / 10);
        curr.next = new ListNode(sum % 10);
        curr = curr.next;
    }

    return dummy.next;
}

// Helper class for testing
class ListNode {
    constructor(val, next) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

// Helper to create list from array
function createList(arr) {
    let dummy = new ListNode(0);
    let curr = dummy;
    for (let val of arr) {
        curr.next = new ListNode(val);
        curr = curr.next;
    }
    return dummy.next;
}

// Helper to print list
function printList(node) {
    let output = [];
    while (node) {
        output.push(node.val);
        node = node.next;
    }
    console.log(output);
}

// Test cases
let l1 = createList([2, 4, 3]);
let l2 = createList([5, 6, 4]);
printList(addTwoNumbers(l1, l2)); // [7,0,8]

module.exports = { addTwoNumbers, ListNode };
